Implement DatetimeArray._from_sequence

jbrockmendel opened the pull request:
Removes dependence of `DatetimeArray.__new__` on `DatetimeIndex`.  De-duplicated `DatetimeIndex.__new__`/`DatetimeArray.__new__`.

The contents of `DatetimeArray._from_sequence` are basically just moved from `DatetimeIndex.__new__`.  This is feasible because #23675 disentangled `to_datetime` from `DatetimeIndex.__new__`.

cc @TomAugspurger this is the last thing on my todo list for DTA/TDA.  LMK if I can be helpful with the composition transition.

jbrockmendel committed:
Implement DatetimeArray._from_sequence

pep8speaks commented:
Hello @jbrockmendel! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/arrays/datetimes.py`](https://github.com/jbrockmendel/pandas/blob/be745aa42ed47a03d98b34947b67c08322da96c1/pandas/core/arrays/datetimes.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/datetimes.py`](https://github.com/jbrockmendel/pandas/blob/be745aa42ed47a03d98b34947b67c08322da96c1/pandas/core/indexes/datetimes.py) !

TomAugspurger commented:
> LMK if I can be helpful with the composition transition.

Mind if I ping you on specific xfails that I've added in that PR after this, 
https://github.com/pandas-dev/pandas/pull/23601, and 
https://github.com/pandas-dev/pandas/pull/23990 are merged? I'm starting to whittle them down.

jbrockmendel commented on pandas/core/arrays/datetimes.py:
```diff
@@ -206,45 +206,103 @@ def _simple_new(cls, values, freq=None, tz=None):
+        if freq is None and hasattr(data, "freq"):
+            # i.e. DatetimeArray/Index
+            freq = data.freq
+            verify_integrity = False
```
I think the thing to do here to to move the ` if freq is None and hasattr(value, "freq"):` check a few lines down so that Series gets unpacked first.

jbrockmendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -49,12 +44,17 @@ def _new_DatetimeIndex(cls, d):
-        #  go through _simple_new instead
-        warnings.simplefilter("ignore")
-        result = cls.__new__(cls, verify_integrity=False, **d)
+    if "data" in d and not isinstance(d["data"], DatetimeIndex):
```
Without this we get a couple of failing pickle-based tests.  tests.frame.test_block_internals.test_pickle and tests.series.test_timeseries.test_pickle

TomAugspurger replied:
Gotcha, I've been fighting those too errors too. I suspect they were working before because DatetimeIndex accepts `data=None` for range-based constructors, which we don't want for the arrays.

jreback replied:
this is because you need to define ``__reduce__`` on in ``ExtensionArray``. This should *never* be hit by a non-DTI.

jreback replied:
I think you need to add back verfiy_integrity to the constructor

jbrockmendel replied:
I think that cases where validate_frequency raises are asking for trouble and the tests cases that fail this check are wonky.

@TomAugspurger you said you've battled with this.  Any thoughts?

jbrockmendel replied:
> and the tests cases that fail this check are wonky.

Yah, tracking one of these down, we are pickling+unpickling `df` defined:

```
@pytest.fixture
def timezone_frame()
    df = DataFrame({'A': date_range('20130101', periods=3),
                    'B': date_range('20130101', periods=3,
                                    tz='US/Eastern'),
                    'C': date_range('20130101', periods=3,
                                    tz='CET')})
    df.iloc[1, 1] = NaT
    df.iloc[1, 2] = NaT
    return df
```

The problem derives from the fact that we've called `__setitem__` (or an equivalent) and should be setting `freq=None` when we do so.

```
>>> dti = df['B']._values
>>> dti
DatetimeIndex(['2013-01-01 00:00:00-05:00', 'NaT',
               '2013-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

>>> DatetimeIndex._validate_frequency(dti, freq=dti.freq)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/pandas/core/indexes/datetimelike.py", line 230, in _validate_frequency
    raise ValueError(msg.format(infer=inferred, passed=freq.freqstr))
ValueError: Inferred frequency None from passed values does not conform to passed frequency D
```

TomAugspurger replied:
> @TomAugspurger you said you've battled with this. Any thoughts?
 
Medium-term, we can change `DatetimeArray.__new__` to `__init__` and all these problems go away (I think). We *may* not be able to do that until after the split though... I haven't tried.

TomAugspurger replied:
I assume Index subclasses are forced to use `__new__` since `Index.__new__` may return subclasses. But our arrays don't have that restriction.

jbrockmendel replied:
This is fixed by #24096

jreback replied:
let's fix that one first then. this needs to be changed here.

TomAugspurger replied:
The changes over in #24096 seem to be... different? I don't know how to explain it, but doesn't the fact that we're having to copy data over in #24096 seem disconnected from pickling?

jbrockmendel replied:
> but doesn't the fact that we're having to copy data over in #24096 seem disconnected from pickling?

Pickling turns out to be only tangentially related to the "real" problem.  In the status quo, altering a datetime64tz column alters the DatetimeIndex that backs it, but doesn't set its `freq` to `None`.  When that DataFrame is pickled and then unpickled, it tries to reconstruct that `DatetimeIndex`, but is passing arguments that _should_ raise `ValueError`.  ATM that gets side-stepped by passing `verify_integrity=False`.

So the goal of #24096 is to not corrupt the DatetimeIndex in the first place, making `verify_integrity=False` unnecessary.

That's still pretty roundabout.  Any clearer?

TomAugspurger replied:
Yeah, that does help.

jbrockmendel commented:
> Mind if I ping you on specific xfails that I've added 

Sounds good.

jbrockmendel committed:
move check down

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -206,45 +206,103 @@ def _simple_new(cls, values, freq=None, tz=None):
+            # assume this data are epoch timestamps
+            if data.dtype != _INT64_DTYPE:
+                data = data.astype(np.int64, copy=False)
+            subarr = data.view(_NS_DTYPE)
```
might be able to pull out
``subarr = data.view(_NS_DTYPE)`` of the if/else and just do it always (or maybe just do it in _simple_new), but this is for later

jbrockmendel replied:
That's what I tried originally, but it breaks the recently-implemented tests.arrays.test_datetimelike.test_from_array_keeps_base (#23956)

TomAugspurger replied:
Hrm sorry. FWIW, that was a hackish thing that I introduced to avoid segfaults. If the code in reduction code was doing the right thing, we may be able to remove that test / requirement. But I'm not sure what the right thing is.

jbrockmendel replied:
For the purposes of this PR it sounds like the options are to either remove that test and un-indent t his line, or keep this line how it is.  It sounds like the first option would cause problems in your DTA branch.  I don't have a strong preference here.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -206,45 +206,103 @@ def _simple_new(cls, values, freq=None, tz=None):
-        if freq is None and hasattr(values, "freq"):
-            # i.e. DatetimeArray, DatetimeIndex
-            freq = values.freq
+    @classmethod
```
is there a reason you don't want to add ``verify_integrity`` here (as maybe ``_verify_integrity=True``)?

jbrockmendel replied:
We've deprecated the kwarg in the DatetimeIndex constructor to get rid of it.  In cases where verify_integrity is not needed, a different constructor (e.g. simple_new) should be used.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -206,45 +206,103 @@ def _simple_new(cls, values, freq=None, tz=None):
+
+        result = cls._simple_new(subarr, freq=freq, tz=tz)
+        if dtype is not None:
+            # TODO: can we work this into validate_tz_from_dtype?
```
should this not be before you call ``_simple_new``?

jbrockmendel commented:
> and the positional arguments name is different

definitely +1 on having these match

> What do you think about moving this PR's current _from_sequence to a different name (not sure what to call it, maybe a top-level datetime_array to mirror period_array?

I'm on board with separating _most_ of the method out into a `sequence_to_dt64ns` mirroring the `sequence_to_td64ns` we have in timedeltas.

I still think having `period_array` instead of having the `PeriodArray` constructor be user-friendly is silly.  I've made my peace with not getting my way on this one, but draw the line at implementing it myself.  It will be easy for e.g. Joris to "fix" these in a follow-up.

(plus there's also `to_datetime`.  -1 on proliferation when `DatetimeArray.__init__` is an obvious fit and `DatetimeArray._from_sequence` also exists)

`sequence_to_dt64ns` notwithstanding, I think the main outstanding concern for this PR is the pickle/verify_integrity thing, which should be handled by either a) @jreback being OK with the change to `_new_DatetimeIndex` implemented here or b) #24096 getting sorted out.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into from_sequence

TomAugspurger commented:
That's all totally fair.

> I think the main outstanding concern for this PR is the pickle/verify_integrity thing, which should be handled by either a) @jreback being OK with the change to _new_DatetimeIndex implemented here or b) #24096 getting sorted out

Of those two, I'm not sure which is preferred. Some scattered observations

1. if the only thing that failed on #24096 was the recent test from https://github.com/pandas-dev/pandas/pull/23956, then feel free to ignore / break that. I suspect that the groupby issue segfaulting there isn't properly fixed yet, but...
2. I think that *all* of these pickle concerns are going to go away once we just use `__init__` instead of `__new__`. It may be worth trying to see if we can do that before #24024.

jbrockmendel commented:
I'm about to push a new commit, the relevant changes being:
- updated `TimedeltaArray._from_sequence` signature to have positional arguments match `ExtensionArray._from_sequence`
- separated most of `DatetimeArray._from_sequence` into `sequence_to_dt64ns`, mirroring timedelta version `sequence_to_td64ns`
- moved `validate_tz_from_dtype` from `arrays.datetimelike` to `arrays.datetimes`, since it is only used there
- added to `validate_tz_from_dtype` a check for tz-naive dtype and non-None tz, which lets us move that particular check to before the call to `_simple_new`
- implemented `datetimelike.validate_inferred_freq` to share some more validation code between `TimedeltaArray._from_sequence` and `DatetimeArray._from_sequence`

`sequence_to_dt64ns` still needs a docstring

jbrockmendel committed:
separate most of from_sequence into sequence_to_dt64ns

jbrockmendel commented:
> if the only thing that failed on #24096 was the recent test from #23956,

That test failed here, not in #24096.  It was fixed by avoiding calling `data = data.view(_NS_DTYPE)` in the case where `data.dtype` is already `M8[ns]`.  The difference in the code amounts to one indentation.  Not worth worrying about.

> I think that all of these pickle concerns are going to go away once we just use __init__ instead of __new__. It may be worth trying to see if we can do that before #24024.

I don't see how that will work, but pickling is an area where I frequently find new and exciting ways to screw up.  My preference would be to merge this with the small `_new_DatetimeIndex` edits and revisit it after #24024.  This would make #24096 not-a-blocker and let me go through those constructors a little more thoroughly (motivated #24100)

jbrockmendel committed:
isort fixup

jbrockmendel committed:
docstring

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1778,3 +1872,52 @@ def _maybe_localize_point(ts, is_none, is_not_none, freq, tz):
     if is_none is None and is_not_none is not None:
         ts = ts.tz_localize(**localize_args)
     return ts
+
```
I would move this up to the section where you the 'Construction Helpers', e.g. the maybe_* routines. (and clarify that name of the section, possibly calling it Validation and Inference )

jbrockmendel replied:
sure.  there's also a one func in datetimelike that is only used by PeriodArray I'd like to move.  might as well do that now

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into from_sequence

jbrockmendel committed:
requested rearrangement

jreback commented:
ok, rebase just in case here

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into from_sequence

jbrockmendel commented:
Rebased.  Haven't reverted the _new_DatetimeIndex change because there is one other case that still fails: there is an overflow in _generate_range that we're not handling correctly.  I'll address that in a separate PR before long.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1444,81 +1428,108 @@ def to_julian_date(self):
     Raises
     ------
-    TypeError : if both timezones are present but do not match
+    TypeError : PeriodDType data is passed
```
is this explicity handled?

jbrockmendel replied:
Via maybe_convert_dtype

jreback commented:
thanks @jbrockmendel merging to unblock things. but as they say in boxing: keep it clean! hahah
